#' Plot gistic results as a bubble plot or as a genomic segment plot.
#' @description Plots significantly altered cytobands as a function of number samples in which it is altered and number genes it contains. Size of each bubble is according to -log10 transformed q values.
#' Optionally if scores.gistic file is provided, a genomic plot is generated with segments highlighting signififcant Amplifications and Deletion regions.
#' @param gistic an object of class \code{GISTIC} generated by \code{readGistic}
#' @param gis.scores optional scores.gistic file generated by GISTIC. If provided plot would be a histogram of amplifications and deletions along the genome. Default NULL
#' @param fdrCutOff fdr cutoff to use. Default 0.1
#' @param markBands any cytobands to label.
#' @param color colors for Amp and Del events.
#' @param ref.build reference build. Could be hg18, hg19 or hg38.
#' @param cytobandOffset if scores.gistic file is given use this to adjust cytoband size.
#' @param file if given saves plot as a pdf.
#' @param width width of the file to be saved.
#' @param height height of the file to be saved.
#' @param txtSize label size for bubbles.
#' @examples
#' all.lesions <- system.file("extdata", "all_lesions.conf_99.txt", package = "maftools")
#' amp.genes <- system.file("extdata", "amp_genes.conf_99.txt", package = "maftools")
#' del.genes <- system.file("extdata", "del_genes.conf_99.txt", package = "maftools")
#' laml.gistic = readGistic(gisticAllLesionsFile = all.lesions, gisticAmpGenesFile = amp.genes, gisticDelGenesFile = del.genes, isTCGA = TRUE)
#' plotGisticResults(gistic = laml.gistic)
#' @return nothing
#' @export

plotGisticResults = function(gistic = NULL, gis.scores = NULL, fdrCutOff = 0.1, markBands = NULL, color = NULL, ref.build = 'hg19', cytobandOffset= 0.03, file = NULL, width = 6, height = 5, txtSize = 3){
  
  if(is.null(gistic) & is.null(gis.scores)){
    stop("Provide either gistic object or scores.gistic file generated by GISTIC")
  }
  
  if(!is.null(gistic)){
    g = getCytobandSummary(gistic)
    g = g[qvalues < fdrCutOff]
    g[,Chromosome := sapply(strsplit(x = g$Wide_Peak_Limits, split = ':'), '[', 1)]
    g[,loc := sapply(strsplit(x = g$Wide_Peak_Limits, split = ':'), '[', 2)]
    g[,Start_Position := sapply(strsplit(x = g$loc, split = '-'), '[', 1)]
    g[,End_Position := sapply(strsplit(x = g$loc, split = '-'), '[', 2)]
    g.lin = transformSegments(segmentedData = g[,.(Chromosome, Start_Position, End_Position, qvalues, Cytoband, Variant_Classification)])
  }
  
  if(is.null(color)){
    color = c('Amp' = 'red', 'Del' = 'blue')
  }
  
  if(!is.null(gis.scores)){
    gis.scores = data.table::fread(input = gis.scores)
    colnames(gis.scores) = c('Variant_Classification', 'Chromosome', 'Start_Position', 'End_Position', 'fdr', 'G_Score', 'Avg_amplitude', 'frequency')
    gis.scores = transformSegments(segmentedData = gis.scores, build = ref.build)
    gis.scores$amp = ifelse(test = gis.scores$Variant_Classification == 'Del', yes = -gis.scores$G_Score-cytobandOffset, no = gis.scores$G_Score+cytobandOffset)
    gis.scores$ystart = ifelse(test = gis.scores$Variant_Classification == 'Del', yes = -cytobandOffset, no = cytobandOffset)
    fdrCutOff = -log10(fdrCutOff)
    gis.scores$Variant_Classification = ifelse(test = as.numeric(gis.scores$fdr) > fdrCutOff, yes = gis.scores$Variant_Classification, no = 'neutral')
    gis.scores$Variant_Classification = factor(gis.scores$Variant_Classification, levels = c('neutral', 'Amp', 'Del'))
    
    if(ref.build == 'hg19'){
      chr.lens = c(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663,
                   146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540,
                   102531392, 90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566,
                   155270560, 59373566)
    } else if(ref.build == 'hg18'){
      chr.lens = c(247249719, 242951149, 199501827, 191273063, 180857866, 170899992,
                   158821424, 146274826, 140273252, 135374737, 134452384, 132349534,
                   114142980, 106368585, 100338915, 88827254, 78774742, 76117153,
                   63811651, 62435964, 46944323, 49691432, 154913754, 57772954)
    } else if(ref.build == 'hg38'){
      chr.lens = c(248956422, 242193529, 198295559, 190214555, 181538259, 170805979,
                   159345973, 145138636, 138394717, 133797422, 135086622, 133275309,
                   114364328, 107043718, 101991189, 90338345, 83257441, 80373285,
                   58617616, 64444167, 46709983, 50818468, 156040895, 57227415)
    }else{
      stop("ref.build can only be hg18, hg19 or hg38")
    }
    
    chr.lens.cumsum = cumsum(chr.lens)
    nchrs = length(unique(gis.scores$Chromosome))
    chr.labels= c(1:22, 'X', 'Y')
    chr.tbl = data.table::data.table(chr = chr.labels, start = c(1, chr.lens.cumsum[1:length(chr.lens.cumsum)-1]), end = chr.lens.cumsum)
    chr.tbl$color = c('black', 'white')
    
    gist.gg = ggplot(data = gis.scores[Variant_Classification %in% 'neutral'], aes(x = Start_Position_updated, xend = End_Position_updated, y= ystart , yend = amp, color = Variant_Classification))+
      geom_segment(alpha = 0.6)+geom_segment(data = gis.scores[!Variant_Classification %in% 'neutral'], alpha = 0.6)+cowplot::theme_cowplot()+theme(legend.position = 'none')+
      scale_color_manual(values = c('Amp'  ='red', 'Del' = 'blue', 'neutral' = 'gray70'))+ylab('G Score')+
      theme(axis.line.x = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank())+xlab("")+
      geom_rect(data = chr.tbl, aes(xmin = start, xmax = end, ymin = -cytobandOffset, ymax = cytobandOffset), fill = chr.tbl$color, inherit.aes = FALSE, alpha = 0.9)
    # scale_x_continuous(breaks = chr.lens.sumsum[1:nchrs], labels = chr.labels[1:nchrs])+
    # geom_vline(xintercept = chr.lens.sumsum[1:nchrs], linetype = 'dotted', size = 0.3, alpha = 0.4)+
    
    if(!is.null(markBands)){
      data.table::setkey(x = g.lin,  Chromosome, Start_Position, End_Position)
      gis.scores = data.table::foverlaps(x = gis.scores, y = g.lin, by.x = c('Chromosome', 'Start_Position', 'End_Position'), type = 'within', mult = 'first')
      gis.scores = gis.scores[order(abs(amp), decreasing = TRUE)]
      g.labs = gis.scores[!duplicated(Cytoband)]
      g.labs = g.labs[Cytoband %like% markBands]
      
      if(nrow(g.labs) == 0){
        message("Nothing to label!")
      }else{
        #nudgeY = max(gis.scores[,abs(G_Score)], na.rm = TRUE)
        #g.labs$nudgeY = ifelse(test = g.labs$Variant_Classification == 'Del', yes = -nudgeY, no = nudgeY)
        gist.gg = gist.gg+ggrepel::geom_text_repel(data = g.labs, aes(x = Start_Position_updated, y = amp, label = Cytoband), size = txtSize, color = 'black', nudge_x = 200e6)
      }
    }
    
  } else{
    
    g$lab = sapply(strsplit(x = g$Unique_Name, split = ':'), '[', 2)
    
    g$pos = ifelse(test = g$Variant_Classification %in% 'Amp', yes = g$nGenes, no = -1 * g$nGenes)
    
    gist.gg = ggplot(data = g, aes(x = nSamples, y = pos, size = -log10(qvalues), color = Variant_Classification, label = lab))+
      geom_point(alpha = 0.6)+xlab('nSamples')+ylab('nGenes')+
      scale_colour_manual(values = color, name = c('CNV', ''))+
      cowplot::theme_cowplot(font_size = 10)+cowplot::background_grid(major = 'xy')+theme(legend.position = 'bottom')
    
    if(!is.null(markBands)){
      g.labs = g[Cytoband %like% markBands]
      if(nrow(g.labs) == 0){
        message("Nothing to label!")
      }else{
        gist.gg = gist.gg+ggrepel::geom_text_repel(data = g.labs, size = txtSize)
      }
    }else{
      gist.gg = gist.gg+ggrepel::geom_text_repel(size = txtSize)
    }
  }
  
  if(!is.null(file)){
    cowplot::save_plot(filename = paste(file, 'pdf', sep='.'), plot = gist.gg, base_height = height, base_width = width)
  }else{
    print(gist.gg)
  }
  
  return(gist.gg)
}
